dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,758 @@
+\
+        getItemsImpl<TYPE, TYPE>( \
+            ind, key_columns, [&](const size_t row, const auto value) { out[row] = value; }, [&](const size_t) { return def; }); \
+    }
+    DECLARE(UInt8)
+    DECLARE(UInt16)
+    DECLARE(UInt32)
+    DECLARE(UInt64)
+    DECLARE(UInt128)
+    DECLARE(Int8)
+    DECLARE(Int16)
+    DECLARE(Int32)
+    DECLARE(Int64)
+    DECLARE(Float32)
+    DECLARE(Float64)
+    DECLARE(Decimal32)
+    DECLARE(Decimal64)
+    DECLARE(Decimal128)
+#undef DECLARE
+
+void IPolygonDictionary::getString(
+        const std::string & attribute_name,
+        const Columns & key_columns,
+        const DataTypes &,
+        const String & def,
+        ColumnString * const out) const
+{
+    const auto ind = getAttributeIndex(attribute_name);
+    checkAttributeType(name, attribute_name, dict_struct.attributes[ind].underlying_type, AttributeUnderlyingType::utString);
+
+    getItemsImpl<String, StringRef>(
+            ind,
+            key_columns,
+            [&](const size_t, const StringRef value) { out->insertData(value.data, value.size); },
+            [&](const size_t) { return StringRef{def}; });
+}
+
+template <typename AttributeType, typename OutputType, typename ValueSetter, typename DefaultGetter>
+void IPolygonDictionary::getItemsImpl(
+        size_t attribute_ind, const Columns & key_columns, ValueSetter && set_value, DefaultGetter && get_default) const
+{
+    const auto points = extractPoints(key_columns);
+
+    using ColVecType = std::conditional_t<IsDecimalNumber<AttributeType>, ColumnDecimal<AttributeType>, ColumnVector<AttributeType>>;
+    using ColType = std::conditional_t<std::is_same<AttributeType, String>::value, ColumnString, ColVecType>;
+    const auto column = typeid_cast<const ColType *>(attributes[attribute_ind].get());
+    if (!column)
+        throw Exception{"An attribute should be a column of its type", ErrorCodes::BAD_ARGUMENTS};
+    for (const auto i : ext::range(0, points.size()))
+    {
+        size_t id = 0;
+        const auto found = find(points[i], id);
+        id = ids[id];
+        if (!found)
+        {
+            set_value(i, static_cast<OutputType>(get_default(i)));
+            continue;
+        }
+        if constexpr (std::is_same<AttributeType, String>::value)
+            set_value(i, static_cast<OutputType>(column->getDataAt(id)));
+        else
+            set_value(i, static_cast<OutputType>(column->getElement(id)));
+    }
+
+    query_count.fetch_add(points.size(), std::memory_order_relaxed);
+}
+
+namespace
+{
+
+inline void makeDifferences(IColumn::Offsets & values)
+{
+    for (size_t i = 1; i < values.size(); ++i)
+        values[i] -= values[i - 1];
+}
+
+struct Offset
+{
+    Offset() = default;
+    IColumn::Offsets ring_sizes, polygon_sizes, multi_polygon_sizes;
+    IColumn::Offset  current_ring = 0, current_polygon = 0, current_multi_polygon = 0;
```

Comment:
I think it would be better to separate this offset structure and iteration counter.
It's also may be more readable to make special Iterator class (iteration code now is smudged now between several functions and hard to follow).

Actually, I think about something like:
```c++
struct Offset
{
    IColumn::Offsets multi_polygon_offsets;
    IColumn::Offsets polygon_offsets;
    IColumn::Offsets ring_offsets;
    
    struct Iterator
    {
        IColumn::Offset current_multi_polygon = 0;
        IColumn::Offset current_polygon = 0;
        IColumn::Offset current_ring = 0;
        IColumn::Offset current_point = 0;

        void next()
        {
            ++current_point;

            if (current_point < ring_offsets[current_ring])
                break;
            
            ++current_ring;

            if (current_ring < polygon_offsets[current_polygon])
                break;

            ++current_polygon;

            if (current_polygon < multi_polygon_offsets[current_multi_polygon])
                break;

            ++current_multi_polygon;
        }

        bool isNewRingStarted() { return current_point == ring_offsets[current_ring - 1]; }
        bool isNewPolygonStarted() { return current_ring == polygon_offsets[current_polygon - 1] && isNewRingStarted(); }
        bool isNewMultiPoligonStarted() { return current_polygon == multi_polygon_offsets[current_multi_polygon - 1] && isNewPolygonStarted(); }
        bool isFinished() { return current_multi_polygon == multi_polygon_offsets.size(); }
    };
};
```
It's a pseudo-code and may be not exactly correct. I also use offsets, not sizes.

Reply:
Github shows this comment as outdated, but it isn't.

Reply:
or you can write methods like `isLastPointOfRing` or `isLastRingOfPolygon` or so on.

Reply:
Tried implementing the approach you suggested, see latest commits.
I would appreciate if you took a look when you get the time.